File: src/test/java/network/crypta/support/HexUtilTest.java
package network.crypta.support;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.math.BigInteger;
import java.util.Arrays;
import java.util.BitSet;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Test case for {@link HexUtil} class.
 *
 * @author Alberto Bacchelli &lt;sback@freenetproject.org&gt;
 */
@SuppressWarnings("java:S100")
class HexUtilTest {

  @Test
  void bytesToHex_whenSingleByteValues_expectTwoDigitHex() {
    // Arrange
    byte[] methodByteArray = new byte[1];

    // Act + Assert
    for (int i = 255; i >= 0; i--) {
      methodByteArray[0] = (byte) i;
      String actual = HexUtil.bytesToHex(methodByteArray);
      String expected = twoDigitHex(i);
      assertEquals(expected, actual);
    }
  }

  @Test
  void hexToBytes_whenSingleByteHexStrings_expectMatchingByteArray() {
    // Arrange
    byte[] expectedByteArray = new byte[1];

    // Act + Assert
    for (int i = 255; i >= 0; i--) {
      expectedByteArray[0] = (byte) i;
      String methodHexString = twoDigitHex(i);
      assertArrayEquals(expectedByteArray, HexUtil.hexToBytes(methodHexString));
    }
  }

  @Test
  void hexToBytesWithOffset_whenOffsetIsZero_expectMatchingByteArray() {
    // Arrange
    byte[] expectedByteArray = new byte[1];

    // Act + Assert
    for (int i = 255; i >= 0; i--) {
      expectedByteArray[0] = (byte) i;
      String methodHexString = twoDigitHex(i);
      assertArrayEquals(expectedByteArray, HexUtil.hexToBytes(methodHexString, 0));
    }
  }

  @Test
  void hexToBytesIntoBuffer_whenSingleByteHexStrings_expectWrittenByte() {
    // Arrange
    byte[] expectedByteArray = new byte[1];
    byte[] outputArray = new byte[1];

    // Act + Assert
    for (int i = 255; i >= 0; i--) {
      expectedByteArray[0] = (byte) i;
      String methodHexString = twoDigitHex(i);
      HexUtil.hexToBytes(methodHexString, outputArray, 0);
      assertArrayEquals(expectedByteArray, outputArray);
    }
  }

  @Test
  void bitsToBytes_whenAllSingleByteBitPatterns_expectSequentialByteValues() {
    // Arrange
    byte[] expectedByteArray = new byte[1];
    BitSet methodBitSet = new BitSet(8);

    // Act + Assert
    for (int i = 0; i < 256; i++) {
      byte[] outputArray = HexUtil.bitsToBytes(methodBitSet, 8);
      expectedByteArray[0] = (byte) i;
      assertArrayEquals(expectedByteArray, outputArray);
      addOne(methodBitSet);
    }
  }

  @Test
  void countBytesForBits_whenValuesCrossByteBoundaries_expectExpectedCounts() {
    // Arrange + Act + Assert
    assertEquals(0, HexUtil.countBytesForBits(0));
    for (int expectedBytesCount = 1; expectedBytesCount < 256; expectedBytesCount++) {
      for (int bits = (expectedBytesCount - 1) * 8 + 1; bits <= expectedBytesCount * 8; bits++) {
        assertEquals(expectedBytesCount, HexUtil.countBytesForBits(bits));
      }
    }
  }

  @Test
  void bytesToBits_whenConvertingSingleByteValues_expectRoundTripViaBitsToBytes() {
    // Arrange
    byte[] methodByteArray = new byte[1];
    BitSet methodBitSet = new BitSet(8);

    // Act + Assert
    for (int i = 0; i < 255; i++) {
      methodByteArray[0] = (byte) i;
      HexUtil.bytesToBits(methodByteArray, methodBitSet, 7);
      assertArrayEquals(methodByteArray, HexUtil.bitsToBytes(methodBitSet, 8));
    }
  }

  @Test
  void biToHex_whenGivenKnownValues_expectExpectedHexStrings() {
    // Arrange + Act + Assert
    BigInteger methodBigInteger = new BigInteger("999999999999999");
    assertEquals("038d7ea4c67fff", HexUtil.biToHex(methodBigInteger));

    methodBigInteger = BigInteger.ZERO;
    assertEquals("00", HexUtil.biToHex(methodBigInteger));

    methodBigInteger = new BigInteger("72057594037927935");
    assertEquals("00ffffffffffffff", HexUtil.biToHex(methodBigInteger));
  }

  @Test
  void bitsToHexString_whenBitsSet_expectExpectedHexRepresentation() {
    // Arrange
    BitSet methodBitSet = new BitSet(8);

    // Act + Assert
    assertEquals("00", HexUtil.bitsToHexString(methodBitSet, 8));

    methodBitSet.set(0, 7, true);
    assertEquals("7f", HexUtil.bitsToHexString(methodBitSet, 8));

    methodBitSet.set(0, 9, true);
    assertEquals("ff", HexUtil.bitsToHexString(methodBitSet, 8));
  }

  @Test
  void hexToBits_whenKnownHexValues_expectBitSetIntersectsExpectedBits() {
    // Arrange
    BitSet expectedBitSet = new BitSet(8);

    // Act
    BitSet bitSetFromZero = new BitSet(8);
    HexUtil.hexToBits("00", bitSetFromZero, bitSetFromZero.size());

    expectedBitSet.set(0, 7, true);
    BitSet bitSetFrom7f = new BitSet(8);
    HexUtil.hexToBits("7f", bitSetFrom7f, bitSetFrom7f.size());

    expectedBitSet.set(0, 9, true);
    BitSet bitSetFromFf = new BitSet(8);
    HexUtil.hexToBits("ff", bitSetFromFf, bitSetFromFf.size());

    // Assert
    assertEquals(0, bitSetFromZero.cardinality());
    assertTrue(bitSetFrom7f.intersects(expectedBitSet));
    assertTrue(bitSetFromFf.intersects(expectedBitSet));
  }

  @Test
  void writeBigIntegerAndReadBigInteger_whenRoundTrippedThroughDataStreams_expectSameValue() {
    // Arrange
    BigInteger methodBigInteger = new BigInteger("999999999999999");
    ByteArrayOutputStream methodByteArrayOutStream = new ByteArrayOutputStream();
    DataOutputStream methodDataOutStream = new DataOutputStream(methodByteArrayOutStream);

    // Act
    BigInteger actual =
        assertDoesNotThrow(
            () -> {
              HexUtil.writeBigInteger(methodBigInteger, methodDataOutStream);
              ByteArrayInputStream methodByteArrayInStream =
                  new ByteArrayInputStream(methodByteArrayOutStream.toByteArray());
              DataInputStream methodDataInStream = new DataInputStream(methodByteArrayInStream);
              return HexUtil.readBigInteger(methodDataInStream);
            });

    // Assert
    assertEquals(0, methodBigInteger.compareTo(actual));
  }

  @Test
  void bytesToHexWithRange_whenOffsetIsTooLong_expectIllegalArgumentException() {
    // Arrange
    int arrayLength = 3;
    byte[] methodBytesArray = new byte[arrayLength];

    // Act
    IllegalArgumentException exception =
        assertThrows(
            IllegalArgumentException.class,
            () -> HexUtil.bytesToHex(methodBytesArray, arrayLength + 1, 1));

    // Assert
    assertNotNull(exception);
  }

  @Test
  void bytesToHexWithRange_whenLengthIsTooLong_expectIllegalArgumentException() {
    // Arrange
    int arrayLength = 3;
    byte[] methodBytesArray = new byte[arrayLength];

    // Act
    IllegalArgumentException exception =
        assertThrows(
            IllegalArgumentException.class,
            () -> HexUtil.bytesToHex(methodBytesArray, 0, arrayLength + 1));

    // Assert
    assertNotNull(exception);
  }

  @Test
  void bytesToHexWithRange_whenLengthIsZero_expectEmptyString() {
    // Arrange
    int length = 0;
    byte[] methodBytesArray = {1, 2, 3};

    // Act
    String actual = HexUtil.bytesToHex(methodBytesArray, 0, length);

    // Assert
    assertEquals("", actual);
  }

  @Test
  void hexToBytesIntoBuffer_whenOffsetIsTooLong_expectArrayIndexOutOfBoundsException() {
    // Arrange
    String methodString = "0";
    byte[] methodByteArray = new byte[1];

    // Act
    ArrayIndexOutOfBoundsException exception =
        assertThrows(
            ArrayIndexOutOfBoundsException.class,
            () -> HexUtil.hexToBytes(methodString, methodByteArray, methodByteArray.length));

    // Assert
    assertNotNull(exception);
  }

  @Test
  void hexToBytesIntoBuffer_whenOutputBufferIsTooShort_expectIndexOutOfBoundsException() {
    // Arrange
    String methodString = "0000";
    byte[] methodByteArray = new byte[1];

    // Act
    IndexOutOfBoundsException exception =
        assertThrows(
            IndexOutOfBoundsException.class,
            () -> HexUtil.hexToBytes(methodString, methodByteArray, 0));

    // Assert
    assertNotNull(exception);
  }

  @Test
  void hexToBytesIntoBuffer_whenInputContainsBadDigit_expectNumberFormatException() {
    // Arrange
    String methodString = "00%0";
    byte[] methodByteArray = new byte[methodString.length()];

    // Act
    NumberFormatException exception =
        assertThrows(
            NumberFormatException.class,
            () -> HexUtil.hexToBytes(methodString, methodByteArray, 0));

    // Assert
    assertNotNull(exception);
  }

  @Test
  void hexToBytesWithOffset_whenInputContainsBadDigit_expectNumberFormatException() {
    // Arrange
    String methodString = "00%0";

    // Act
    NumberFormatException exception =
        assertThrows(NumberFormatException.class, () -> HexUtil.hexToBytes(methodString, 0));

    // Assert
    assertNotNull(exception);
  }

  @Test
  void hexToBytes_whenInputContainsBadDigit_expectNumberFormatException() {
    // Arrange
    String methodString = "00%0";

    // Act
    NumberFormatException exception =
        assertThrows(NumberFormatException.class, () -> HexUtil.hexToBytes(methodString));

    // Assert
    assertNotNull(exception);
  }

  @Test
  void bitsToBytes_whenSizeIsSmallerThanSetBits_expectTruncatedResult() {
    // Arrange
    byte[] expectedByteArray = new byte[1];
    BitSet methodBitSet = new BitSet(8);
    methodBitSet.flip(0);
    expectedByteArray[0] = (byte) 1;

    // Act
    byte[] outputAtSizeZero = HexUtil.bitsToBytes(methodBitSet, 0);
    byte[] outputAtSizeOne = HexUtil.bitsToBytes(methodBitSet, 1);

    // Assert
    assertFalse(Arrays.equals(expectedByteArray, outputAtSizeZero));
    assertArrayEquals(expectedByteArray, outputAtSizeOne);

    // Arrange
    methodBitSet.flip(7);
    methodBitSet.flip(3);
    expectedByteArray[0] = (byte) (128 + 8 + 1);

    // Act
    byte[] outputAtSizeThree = HexUtil.bitsToBytes(methodBitSet, 3);
    byte[] outputAtSizeEight = HexUtil.bitsToBytes(methodBitSet, 8);

    // Assert
    assertFalse(Arrays.equals(expectedByteArray, outputAtSizeThree));
    assertArrayEquals(expectedByteArray, outputAtSizeEight);
  }

  @Test
  void bytesToHexWithRange_whenLengthZeroAndOffsetAtEnd_expectEmptyString() {
    // Arrange + Act + Assert
    assertEquals("", HexUtil.bytesToHex(new byte[0], 0, 0));
    assertEquals("", HexUtil.bytesToHex(new byte[2], 2, 0));
  }

  private static String twoDigitHex(int value) {
    return value <= 15 ? "0" + Integer.toHexString(value) : Integer.toHexString(value);
  }

  private static void addOne(BitSet aBitSet) {
    int bitSetIndex = 0;
    while (aBitSet.get(bitSetIndex)) {
      aBitSet.flip(bitSetIndex);
      bitSetIndex++;
    }
    aBitSet.flip(bitSetIndex);
  }
}

File: src/test/java/network/crypta/support/HexUtilTest.kt
package network.crypta.support

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.EOFException
import java.io.IOException
import java.math.BigInteger
import java.util.BitSet
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource

@Suppress("java:S100") // Allow expressive test method names
class HexUtilTest {

  // bytesToHex ---------------------------------------------------------------

  @Test
  fun bytesToHex_whenAllSingleBytes_expectTwoDigitLowercase() {
    // Arrange
    val one = ByteArray(1)

    // Act + Assert
    for (i in 0..255) {
      one[0] = i.toByte()
      val expected = Integer.toHexString(i).padStart(2, '0')
      assertEquals(expected, bytesToHex(one))
    }
  }

  @Test
  fun bytesToHex_whenZeroLength_expectEmpty() {
    assertEquals("", bytesToHex(ByteArray(0), 0, 0))
    assertEquals("", bytesToHex(ByteArray(2), 2, 0))
  }

  @Test
  fun bytesToHex_whenOffsetBeyondArray_expectIllegalArgumentException() {
    val arr = ByteArray(3)
    assertThrows(IllegalArgumentException::class.java) { bytesToHex(arr, 4, 1) }
  }

  @Test
  fun bytesToHex_whenReadingTooManyBytes_expectIllegalArgumentException() {
    val arr = ByteArray(3)
    assertThrows(IllegalArgumentException::class.java) { bytesToHex(arr, 0, 4) }
  }

  @Test
  fun bytesToHexAppend_whenAppendsToExistingBuilder_expectBuilderExtended() {
    val bytes = byteArrayOf(0x00, 0x7F, 0xFF.toByte())
    val sb = StringBuilder("prefix-")
    bytesToHexAppend(bytes, 0, bytes.size, sb)
    assertEquals("prefix-007fff", sb.toString())
  }

  // hexToBytes ---------------------------------------------------------------

  @Test
  fun hexToBytes_whenAllSingleBytes_expectDecodeMatches() {
    val expected = ByteArray(1)
    for (i in 0..255) {
      expected[0] = i.toByte()
      val hex = Integer.toHexString(i).padStart(2, '0')
      assertArrayEquals(expected, hexToBytes(hex))
      assertArrayEquals(expected, hexToBytes(hex, 0))

      val out = ByteArray(1)
      hexToBytes(hex, out, 0)
      assertArrayEquals(expected, out)
    }
  }

  @Test
  fun hexToBytes_whenUpperAndLowercase_expectEquivalent() {
    assertArrayEquals(byteArrayOf(0xAF.toByte()), hexToBytes("af"))
    assertArrayEquals(byteArrayOf(0xAF.toByte()), hexToBytes("AF"))
  }

  @Test
  fun hexToBytes_whenOddLength_expectPrefixedZeroNibble() {
    assertArrayEquals(byteArrayOf(0x0F), hexToBytes("f"))
    assertArrayEquals(hexToBytes("0f"), hexToBytes("f"))
  }

  @Test
  fun hexToBytes_withOffsetPrefix_expectLeadingZerosThenDecoded() {
    val out = hexToBytes("a1b2", 2)
    assertArrayEquals(byteArrayOf(0x00, 0x00, 0xA1.toByte(), 0xB2.toByte()), out)
  }

  @Test
  fun hexToBytes_intoProvidedArray_whenArrayTooSmallEvenLength_expectIndexOutOfBounds() {
    val out = ByteArray(1)
    assertThrows(IndexOutOfBoundsException::class.java) { hexToBytes("abcd", out, 0) }
  }

  @Test
  fun hexToBytes_intoProvidedArray_whenArrayTooSmallOddLength_expectIndexOutOfBounds() {
    // Odd-length input hits the implicit prefixing path; ensure bounds are still enforced.
    val out = ByteArray(1)
    assertThrows(IndexOutOfBoundsException::class.java) { hexToBytes("abc", out, 0) }
  }

  @Test
  fun hexToBytes_intoProvidedArray_whenOffsetEqualsLength_expectIndexOutOfBounds() {
    val out = ByteArray(1)
    assertThrows(IndexOutOfBoundsException::class.java) { hexToBytes("0", out, out.size) }
  }

  @Test
  fun hexToBytes_whenInvalidDigit_expectNumberFormatException() {
    val bad = "00%0"
    assertThrows(NumberFormatException::class.java) { hexToBytes(bad) }
    assertThrows(NumberFormatException::class.java) { hexToBytes(bad, 0) }
    assertThrows(NumberFormatException::class.java) { hexToBytes(bad, ByteArray(bad.length), 0) }
  }

  // BitSet/byte conversions --------------------------------------------------

  @Test
  fun bitsToBytes_whenEnumerateAll8BitValues_expectExactByte() {
    val bs = BitSet(8)
    for (i in 0..255) {
      val bytes = bitsToBytes(bs, 8)
      assertArrayEquals(byteArrayOf(i.toByte()), bytes)
      incrementBitSet(bs)
    }
  }

  @Test
  fun bitsToBytes_whenSizeSmallerThanSetBits_expectMasking() {
    val bs = BitSet(8)
    // 0x01
    bs.flip(0)
    assertFalse(bitsToBytes(bs, 0).contentEquals(byteArrayOf(1)))
    assertArrayEquals(byteArrayOf(1), bitsToBytes(bs, 1))

    // 0x89 (bits 0,3,7)
    bs.flip(7)
    bs.flip(3)
    assertFalse(bitsToBytes(bs, 3).contentEquals(byteArrayOf(0x89.toByte())))
    assertArrayEquals(byteArrayOf(0x89.toByte()), bitsToBytes(bs, 8))
  }

  @Test
  fun bytesToBits_thenBack_when8Bits_expectRoundTrip() {
    val bs = BitSet(8)
    val one = ByteArray(1)
    for (i in 0..254) {
      one[0] = i.toByte()
      bs.clear()
      bytesToBits(one, bs, 7)
      assertArrayEquals(one, bitsToBytes(bs, 8))
    }
  }

  @Test
  fun bitsToHexString_whenExamples_expectExpectedHex() {
    val bs = BitSet(8)
    assertEquals("00", bitsToHexString(bs, 8))
    bs.set(0, 7, true) // 0x7f
    assertEquals("7f", bitsToHexString(bs, 8))
    bs.set(0, 9, true) // 0xff when size=8
    assertEquals("ff", bitsToHexString(bs, 8))
  }

  @Test
  fun hexToBits_whenExamples_expectExpectedBitSets() {
    var bs = BitSet(8)
    hexToBits("00", bs, bs.size())
    assertEquals(0, bs.cardinality())

    bs = BitSet(8)
    hexToBits("7f", bs, bs.size())
    val expected7f = BitSet(8).apply { set(0, 7, true) }
    assertTrue(bs.intersects(expected7f))

    bs = BitSet(8)
    hexToBits("ff", bs, bs.size())
    val expectedff = BitSet(8).apply { set(0, 9, true) }
    assertTrue(bs.intersects(expectedff))
  }

  // Count bytes for bits -----------------------------------------------------

  @ParameterizedTest
  @CsvSource(value = ["0,0", "1,1", "7,1", "8,1", "9,2", "15,2", "16,2", "17,3"])
  fun countBytesForBits_whenVariousSizes_expectCeilDivBy8(bits: Int, expectedBytes: Int) {
    assertEquals(expectedBytes, countBytesForBits(bits))
  }

  // BigInteger I/O -----------------------------------------------------------

  @Test
  fun writeBigInteger_and_readBigInteger_whenRoundTrip_expectSameValue() {
    val value = BigInteger("999999999999999")
    val bos = ByteArrayOutputStream()
    DataOutputStream(bos).use { dos -> writeBigInteger(value, dos) }
    val bytes = bos.toByteArray()
    val read = DataInputStream(ByteArrayInputStream(bytes)).use { dis -> readBigInteger(dis) }
    assertEquals(0, value.compareTo(read))
  }

  @Test
  fun writeBigInteger_whenNegative_expectIllegalArgumentException() {
    val neg = BigInteger.valueOf(-1)
    val bos = ByteArrayOutputStream()
    val dos = DataOutputStream(bos)
    assertThrows(IllegalArgumentException::class.java) { writeBigInteger(neg, dos) }
  }

  @Test
  fun writeBigInteger_whenTooLong_expectIllegalStateException() {
    // Construct a positive BigInteger whose two's-complement byte array exceeds Short.MAX_VALUE
    val magnitude = ByteArray(Short.MAX_VALUE.toInt() + 1) { 0x7F.toByte() }
    val big = BigInteger(1, magnitude)
    val dos = DataOutputStream(ByteArrayOutputStream())
    assertThrows(IllegalStateException::class.java) { writeBigInteger(big, dos) }
  }

  @Test
  fun readBigInteger_whenNegativeLengthHeader_expectIOException() {
    val bos = ByteArrayOutputStream()
    DataOutputStream(bos).use { it.writeShort(-1) }
    val dis = DataInputStream(ByteArrayInputStream(bos.toByteArray()))
    assertThrows(IOException::class.java) { readBigInteger(dis) }
  }

  @Test
  fun readBigInteger_whenUnexpectedEOF_expectIOException() {
    val bos = ByteArrayOutputStream()
    DataOutputStream(bos).use {
      it.writeShort(4) // claims 4 bytes follow
      it.write(byteArrayOf(1, 2, 3)) // only 3 bytes provided
    }
    val dis = DataInputStream(ByteArrayInputStream(bos.toByteArray()))
    val ex = assertThrows(IOException::class.java) { readBigInteger(dis) }
    // DataInputStream.readFully throws EOFException specifically
    assertTrue(ex is EOFException)
  }

  // BigInteger → hex ---------------------------------------------------------

  @Test
  fun biToHex_and_toHexString_whenKnownValues_expectExpectedHex() {
    var bi = BigInteger("999999999999999")
    assertEquals("038d7ea4c67fff", biToHex(bi))
    assertEquals(biToHex(bi), toHexString(bi))

    bi = BigInteger.ZERO
    assertEquals("00", biToHex(bi))
    assertEquals(biToHex(bi), toHexString(bi))

    bi = BigInteger("72057594037927935")
    assertEquals("00ffffffffffffff", biToHex(bi))
    assertEquals(biToHex(bi), toHexString(bi))
  }

  // Helpers ------------------------------------------------------------------

  private fun incrementBitSet(bs: BitSet) {
    var idx = 0
    while (bs[idx]) {
      bs.flip(idx)
      idx++
    }
    bs.flip(idx)
  }
}
